core/unix/signal.c

```diff
@@ -4801,7 +4801,15 @@ master_signal_handler_C(byte *xsp)
         return;
     }
 #endif
-    dcontext_t *dcontext = get_thread_private_dcontext();
+    /* We avoid using safe_read_tls_magic during detach. This thread may already have
+     * lost its TLS. A safe read may result into a race affecting asynchronous non-alarm
+     * signals (xref i#3535) between delivering the SIGSEGV and restoring the app's signal
+     * handlers. We don't need the thread's private dcontext anymore here at this point.
```

Comment:
I'm not sure about this: a NULL dcontext for a non-alarm, non-suspend signal is going to raise "ERROR: master_signal_handler with no siginfo " in debug build, and if it's not clearly asynchronous it's going to kill the whole process with FAILED_TO_HANDLE_SIGNAL below.  That does not sound good: so it does not sound like "We don't need the dcontext anymore" to me.

Reply:
This is really just a band-aid rather than a fix. The reason for this to work in release is

/* see FIXME comments above.                                                                                                                                                                                                                                                      
 * workaround for now: suppressing is better than dying.                                                                                                                                                                                                                          
 */
  if (can_always_delay[sig])
     return;

This is really a bigger problem than just this patch (as you prob. know)
The rationale was to let such a signal slip through (i.e. drop) in release, and fail in debug.

Re-reading the comment is indeed misleading, I am re-phrasing with XXX comment. To really fix this, one had to...

Maybe re-direct signals to another thread? What if thread-directed / comment already mentions this?
Maybe remember the signal, finish detach and resend at the very end of detach? Is this possible?

That said, I am fine to drop this patch completely if you want. I do think that dropping may be better than passing SIGSEGV to the app? Both is not ideal.

Reply:
Like I said, for a "not clearly asynchronous" (== !can_always_delay) signal this will crash in release, when it might have worked if there were a dcontext: so it seems like it is not a clear positive change for such signals.

Reply:
I would suggest having a targeted approach that directly does what we want right here, locally, without leaving a NULL dcontext across dozens of lines of code and basically hoping the right thing happens when we get down to code down below, which seems too fragile.

The goal seems to be to drop asynchronous signals.  We already try to drop alarm signals by removing our handler and setting to SIG_IGN, but there can be a race where an alarm makes it to here, so also setting SIG_IGN for all can_always_delay signals won't fully solve this.

What I would do is have a check for `doing_detach && can_always_delay` and just drop the signal right here, right now.  That seems more readable, more robust, simpler, with no dependencies on the code below and no dangling NULL pointer.

The only downside is that it relates to the #26 code below so we may want to tie them together somehow; have the same warning here?  But on detach I think it's ok to not have the warning so maybe just a comment that the two pieces of code are related.

Reply:
Leave error for debug build?